File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <iostream>
#include <stack>

using namespace std;

bool solution(string s)
{
    bool answer = true;

    stack<char> st;
    
    for(int i=0; i<s.size(); i++){
        if(st.empty() || s[i] == '('){
            st.push('(');
        } else if(!st.empty() && st.top() == '('){
            st.pop();
        }
    }
    
    if(!st.empty()) answer = false;

    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;

        ArrayDeque<Character> st = new ArrayDeque<>();
        
        for(int i=0; i<s.length(); i++){
            if(s.charAt(i) == '('){
                st.push(s.charAt(i));
            }
            else{
                if(st.isEmpty() || st.pop() == s.charAt(i))
                    return false;
            }
        }

        return st.isEmpty();
    }
}